fhevm/tee_bit.go

```diff
@@ -0,0 +1,145 @@
+package fhevm
+
+import (
+	"github.com/ethereum/go-ethereum/common"
+	"go.opentelemetry.io/otel/trace"
+)
+
+func teeShlRun(environment EVMEnvironment, caller common.Address, addr common.Address, input []byte, readOnly bool, runSpan trace.Span) ([]byte, error) {
+	return doShiftOperationGeneric(environment, caller, input, runSpan, func(a, b, typ uint64) uint64 {
+		switch typ {
+		case 1:
+			return uint64((uint8(a) << uint8(b)))
+		case 2:
+			return uint64((uint8(a) << uint8(b)))
+		case 3:
+			return uint64((uint16(a) << uint16(b)))
+		case 4:
+			return uint64((uint32(a) << uint32(b)))
+		case 5:
+			return (a << b)
+		default:
+			return 0
+		}
+	}, "teeShlRun")
+}
+
+func teeShrRun(environment EVMEnvironment, caller common.Address, addr common.Address, input []byte, readOnly bool, runSpan trace.Span) ([]byte, error) {
+	return doShiftOperationGeneric(environment, caller, input, runSpan, func(a, b, typ uint64) uint64 {
+		switch typ {
+		case 1:
+			return uint64((uint8(a) >> uint8(b)))
+		case 2:
+			return uint64((uint8(a) >> uint8(b)))
+		case 3:
+			return uint64((uint16(a) >> uint16(b)))
+		case 4:
+			return uint64((uint32(a) >> uint32(b)))
+		case 5:
+			return (a >> b)
+		default:
+			return 0
+		}
+	}, "teeShrRun")
+}
+
+func teeRotlRun(environment EVMEnvironment, caller common.Address, addr common.Address, input []byte, readOnly bool, runSpan trace.Span) ([]byte, error) {
+	return doShiftOperationGeneric(environment, caller, input, runSpan, func(a, b, typ uint64) uint64 {
+		switch typ {
+		case 1:
+			return uint64((uint8(a) << uint8(b)) | (uint8(a) >> (uint8(4) - uint8(b)%4)))
```

Comment:
Do you know if in TFHE it's possible to do a shift or rotate, where the lhs and rhs arguments are not of same type?

Reply:
In shift case, convert to same type. Plz see [this](https://github.com/zama-ai/fhevm/blob/2ceb2bfef1301c0f645b9f4a5c1a18819114b9d8/lib/TFHE.sol#L5083-L5091).

Reply:
In rotate case, rhs type can be only euint8 or uint8. https://github.com/zama-ai/fhevm/blob/2ceb2bfef1301c0f645b9f4a5c1a18819114b9d8/lib/TFHE.sol#L5056-L5061

Reply:
Sounds good, thanks 👍